quinn/transformations.py

```diff
@@ -87,28 +88,193 @@ def to_snake_case(s: str) -> str:
+                    sort_order=sort_order
+                )
             )
-        )
-    return df.select(*sorted_col_names)
+        reverse_lookup = {
+            "asc": False,
+            "desc": True,
+        }
+        return reverse_lookup[sort_order]
+
+    def sort_top_level_cols(df, is_reversed) -> DataFrame:
+        # sort top level columns
+        sorted_col_names = sorted(df.columns, reverse=is_reversed)
+        return df.select(*sorted_col_names)
+
+    def sort_nested_cols(schema, is_reversed, base_field="") -> list[str]:
+        # recursively check nested fields and sort them
+        # https://stackoverflow.com/questions/57821538/how-to-sort-columns-of-nested-structs-alphabetically-in-pyspark
+        # Credits: @pault for logic
+
+        def parse_fields(
+            fields_to_sort: list, parent_struct, is_reversed: bool
+        ) -> list:
+            sorted_fields: list = sorted(
+                fields_to_sort,
+                key=lambda x: x["name"],
+                reverse=is_reversed,
+            )
+
+            results = []
+            for field in sorted_fields:
+                new_struct = StructType([StructField.fromJson(field)])
+                new_base_field = parent_struct.name
+                if base_field:
+                    new_base_field = base_field + "." + new_base_field
+
+                results.extend(
+                    sort_nested_cols(new_struct, is_reversed, base_field=new_base_field)
+                )
+            return results
+
+        def handle_array_type(parent_struct: StructField, is_reversed: bool) -> str:
+            def format_array_selection(
+                elements: list, base_str: str, suffix_str: str
+            ) -> str:
+                element_names = [i.split(".")[-1] for i in elements]
+                array_elements_formatted = [f"x.{i} as {i}" for i in element_names]
+
+                output = (
+                    f"{base_str} {', '.join(array_elements_formatted)} {suffix_str}"
+                )
+                return output
+
+            array_parent = parent_struct.jsonValue()["type"]["elementType"]
+
+            base_str = f"transform({parent_struct.name}"
+            suffix_str = f") AS {parent_struct.name}"
+
+            # if struct in array, create mapping to struct
+            if array_parent["type"] == "struct":
+                array_parent = array_parent["fields"]
+
+                base_str = f"{base_str}, x -> struct("
+                suffix_str = f"){suffix_str}"
+
+            array_elements = parse_fields(array_parent, parent_struct, is_reversed)
+            formatted_array_selection = format_array_selection(
+                array_elements, base_str, suffix_str
+            )
+            return formatted_array_selection
+
+        def handle_struct_type(parent_struct: StructField, is_reversed: bool) -> str:
+            def format_struct_selection(elements: list, struct_name: str) -> str:
```

Comment:
May we refactor it and avoid so much level of nested functions and so high cognitive complexity?